Quasilumin Escort for CoP 8-2

jmcmorris opened the pull request:
**_I affirm:_**
- [x] I have paid attention to this example and will edit again if need be to not break the formatting, or I will be ignored
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md)
- [x] I've _**tested my code and the things my code has changed**_ since the last commit in the PR, and will test after any later commits

## What does this pull request do?

This implements the logic behind escorting the Quasilumin spawned at the Cermet Alcove's in Grand Palace of Hu'Xzoi. This was reasonably tricky to get right due as there are many moving parts including several that were not properly supported. Such as Aggroable Mobs in #3002 and mobs being treated more like NPCs (triggerable).

https://ffxiclopedia.fandom.com/wiki/A_Fate_Decided

## Steps to test these changes

```
!gotoid 16916927
!gotoid 16916928
!gotoid 16916929
!gotoid 16916930
```

Go to any of those Cermet Alcove's and interact with it to spawn the Quasilumin from it. You can trigger the Quasilumin to stop it and trigger it again to resume. Once it gets to the destination you can interact with it to open the final door.

zach2good committed:
[COP] Start porting Eden's 8-2

Co-authored-by: InoUno <inoiown@gmail.com>
Co-authored-by: Jasson McMorris <jmcmorris@users.noreply.github.com>
Co-authored-by: 59blargedy <52636208+59blargedy@users.noreply.github.com>
Co-authored-by: KnowOne134 <35616771+KnowOne134@users.noreply.github.com>
Co-authored-by: Scott Lassen <6099499+slassen@users.noreply.github.com>

CriticalXI committed:
Merge pull request #419 from Scarifi3dXI/staging

Patrol fix, updates to Kuftal NM pathing

(cherry picked from commit f3deecf2279bf410bf2ecfab98ed342225d84599)

Co-authored-by: Scarifi3dXI <112238268+Scarifi3dXI@users.noreply.github.com>

jmcmorris committed:
Adding one way Cermet Portal Security Gates in Grant Palace of Hu'Xzoi

jmcmorris committed:
Add support for mobs on the player allegiance to be aggroed by other mobs

jmcmorris committed:
Make it possible to trigger normal status mobs

jmcmorris committed:
Implement a dynamic entity quasilumin for escorting in CoP 8-2

jmcmorris committed:
Change special spawn animation in entity packet to be specifically the bit needed

claywar committed:
[mob] Increase Ancient Vessel HP to 12000

WinterSolstice8 committed:
[core] Move entity update in SpawnMob/Trust/Pet back inside if condition

claywar committed:
Merge pull request #3011 from LandSandBoat/ancient-vessel

[mob] Increase Ancient Vessel HP to 12000

claywar committed:
Interaction - Quest: Blade of Death

Additional Fix for Blade of Darkness to require prevZone to be Palborough Mines when zoning into Zeruhn Mines

Fix syntax

claywar committed:
Merge pull request #3012 from WinterSolstice8/fix_0x00E_packet_spam

[core] Move entity update in SpawnMob/Trust/Pet back inside if condition

jmcmorris committed:
Allow mobs to properly engage when being hit with a prevent action effect first

claywar committed:
[ci] Require padding between lua operators and comparators


Ignore quoted strings

WinterSolstice8 committed:
[core] Move removing attack & detectable effect flags after auto attack

claywar committed:
Merge pull request #3015 from WinterSolstice8/fix_sneak_attack

Sneak attack fix

claywar committed:
Merge pull request #3013 from LandSandBoat/interaction/blade-of-death

Interaction - Quest: Blade of Death

zach2good commented on scripts/zones/Grand_Palace_of_HuXzoi/mobs/DE_Quasilumin.lua:
```diff
@@ -0,0 +1,307 @@
+    COMPLETE = 3, -- Has reached the end of the escort mission
+}
+
+local this = {}
```
`entity`

zach2good commented on scripts/zones/Grand_Palace_of_HuXzoi/mobs/DE_Quasilumin.lua:
```diff
@@ -0,0 +1,307 @@
+-----------------------------------
```
Can we move the contents of this file into a helper in this zones folder, or into the triggering zone or npc lua?

zach2good replied:
Failing that, some comments explaining this is a dynamic entity and this is the naming convention would be welcome. Its a little arcane (my bad)

jmcmorris replied:
Sure I don't mind doing either of these. Can you explain the motivation so I can understand better? Thanks!

jmcmorris replied:
@zach2good think you could go into some details? :)

zach2good replied:
I don't think the average person poking around in scripting is going to know that these Quasis are DE's, or necessarily what a DE is or does. The name is defined here:

https://github.com/LandSandBoat/server/blob/base/src/map/lua/lua_zone.cpp#L272

I think attaching the logic at the point of creation, or adding some very clear comments about why this file exists and is named the way it is would help out anyone whos poking around in here

jmcmorris replied:
Thanks! 👍

zach2good commented on scripts/zones/Grand_Palace_of_HuXzoi/IDs.lua:
```diff
@@ -20,26 +20,33 @@ zones[xi.zone.GRAND_PALACE_OF_HUXZOI] =
-        [16916904] = 7376,
-        [16916905] = 7377,
-        [16916906] = 7378,
+        [16916897] = 7380,
```
Eventually, we'll need/want to replace all of these hardcoded ids with some lookups. But in the in interest of getting COP 'complete' we can ignore it for now

jmcmorris replied:
They're sequential so we could technically replace it with just one "offset" and calculate it from there. `textID = textOffset + (mob:getID() - quasiluminOffset)`

zach2good replied:
+1 for offset

zach2good replied:
The main fear is: If SE decides to put a JP or exemplar camp in Grand Palace, or does _anything else_ to the entity list, these IDs might shift and all of this might break. So looking up using names or offsets from other mobs is a safer way of handling this.

zach2good commented on scripts/zones/Grand_Palace_of_HuXzoi/mobs/DE_Quasilumin.lua:
```diff
@@ -0,0 +1,307 @@
+        elseif progress == escortProgress.COMPLETE then
+            mob:showText(mob, ID.text.DUTY_COMPLETE)
+            player:messageSpecial(data.direction)
+            -- TODO: display animation and NPC is not pushing update packet to players in range
```
Is this important?

jmcmorris replied:
I _think_ I was the original person who wrote that comment but it was nearly 3 years ago and I honestly don't remember what it is talking about. I thought about deleting it but thought maybe someone else might know haha

zach2good replied:
Worth leaving it then 👍

zach2good reviewed (approved):
Looking good 👍

claywar committed:
Merge pull request #3014 from jmcmorris/engaged_sleep

Mob Engage When Hit With Prevent Action Effect

claywar committed:
Merge pull request #3002 from jmcmorris/aggroable_mobs

Aggroable Mobs

claywar committed:
[chore] Add required padding for operators and comparators

claywar committed:
[ci] Ignore block comments in lua, add padding requirement for `/`

claywar committed:
[chore] Add required padding to division operators

WinterSolstice8 committed:
[core] Fix block rate wrapping with Ochain + Reprisal, fix reprisal spikes not appearing

* Remove flooring from block rate: https://www.bg-wiki.com/ffxi/Shield_Skill

jmcmorris committed:
Log errors that occur in lua actions and fix one such issue

jmcmorris committed:
Remove pet confrontation status effect when player loses theirs

jmcmorris committed:
Remove all battlefield registered players' confrontation status effect when cleaning up as they may have exited the battlefield before it is cleaned up

claywar committed:
Merge pull request #3021 from jmcmorris/action_queue_errors

Log Action Queue Errors

jmcmorris committed:
Add MP to Kaiser Behemoth in Apollyon NW so it can cast Meteor

jmcmorris committed:
Make Kaiser Behemoth cast Meteor all of the time

claywar committed:
Merge pull request #3022 from jmcmorris/battlefield_exit

Better Battlefield Exiting

claywar committed:
Merge pull request #3023 from jmcmorris/kaiser_behemoth_mp

Kaiser Behemoth cast Meteor in Apollyon NW

claywar committed:
Merge pull request #3019 from LandSandBoat/ci-require-operator-padding

[ci] Require padding around comparators and operators

claywar committed:
[ci] Require specific multiline conditional formatting

Allow only the following formats:
```
if <condition> then

if
    <condition> and
    not <condition2>
end
```

This requires every conditional line to either have both if/elseif and then on a line, or that no data is contained on the line at all.

jmcmorris committed:
Add battlefield effect for registrant earlier so that way the effect's subpower in `onBattlefieldEnter` is set to indicate that the registrant is inside the battlefield so they'll be properly ejected if they log back in after the battlefield has been cleaned up

claywar committed:
[chore] Update multiline conditionals to conform with CI

jmcmorris committed:
Add MOBMOD_DETECTION to support overriding detection method for individual mobs

jmcmorris committed:
Make Bialozar and Thiazi in Apollyon NE detect by sound

jmcmorris committed:
Make Bialozar and Thiazi immune to sleep

WinterSolstice8 committed:
[core] Clear old traits on mob spawn

WinterSolstice8 commented on src/map/packets/entity_update.cpp:
```diff
@@ -82,7 +82,7 @@ void CEntityUpdatePacket::updateWith(CBaseEntity* PEntity, ENTITYUPDATE type, ui
             if (PEntity->spawnAnimation == SPAWN_ANIMATION::SPECIAL)
             {
-                ref<uint8>(0x28) |= 0x45;
+                ref<uint8>(0x28) |= 0x04;
```
0x04 for Quasilumins I presume?

This is an interesting thing that needs to be configurable. Currently, Dragoon Wyverns always use the same "come from the sky" animation upon spawning in range (at 50'+) even if it's not their initial spawn. The intent of this "SPAWN_ANIMATION::SPECIAL" currently was to _specifically_ override the spawn animation for dynamic entities.

I know it's out of scope of this PR, but it might be good to convert this to a mobmod so we can use any arbitrary uint16 value for spawns to accomplish animations that may change.

I could make a separate PR to add that functionality to make Dragoon's Wyverns perform properly which will also allow you to select the proper spawn animation for what you need here, too.

TeoTwawki replied:
many entities do this with different animations. even npcs - the synergy furnaces have a spawn in animation that is different from the standard fade in even, tho you usually will not get close enough to them in time to notice it unless you lag a bit.

There are possibly enough entities needing a non default one that it may eventually need to be stored in the db columns, (let us hope not) but a mobmob would be a good start for now.

jmcmorris replied:
The entire `0x28` bitfield needs to be figured out but I'm fairly sure that `0x04` bit is more or less "play special animation on spawn".

TeoTwawki replied:
oh ok so it's not that multiple animations can fill the byte (so far as we know) it's a flag that uses whatever alternate the model has baked in. that makes sense.

jmcmorris replied:
I was able to figure out a few of the bits such:

- `0x01` seems to prevent entity trigger even on green name (maybe something else but this was a side-effect)
- `0x04` play special spawn (and other?) animations
- `0x40` make the entity triggerable on client

jmcmorris replied:
We have a few others in the code already too:

- ref<uint8>(0x28) |= PMob->StatusEffectContainer->HasStatusEffect(EFFECT_TERROR) ? 0x10 : 0x00;
- ref<uint8>(0x28) |= PMob->health.hp > 0 && PMob->animation == ANIMATION_DEATH ? 0x08 : 0;

WinterSolstice8 replied:
It doesn't seem that the change here affects dynamic entities in any way. Tested with NPC module + !fafnir

claywar committed:
Merge pull request #3028 from jmcmorris/mobmod_detection

Mob modifier for detection

claywar committed:
Merge pull request #3029 from WinterSolstice8/mob_respawn_trait_fix

[core] Clear old traits on mob spawn

claywar committed:
Merge pull request #3020 from WinterSolstice8/fix_reprisal

[core] Fix block rate wrapping with Ochain + Reprisal, fix reprisal spikes not appearing

claywar committed:
Merge pull request #3025 from jmcmorris/battlefield_ejection

Battlefield Ejection For Registrant

claywar committed:
Merge pull request #3026 from LandSandBoat/ci-multiline-conditionals

[ci] Require multiline conditionals to not have data on start or end of blocks

jmcmorris committed:
Bringing over many NPC pathing scripts from ASB that were done by Ophelia and updated by me

Co-authored-by: OpheliaXI <heliashelium@gmail.com>

jmcmorris committed:
Rename local path variable to pathNodes to help distinguish it from xi.path per Teo's request

TeoTwawki committed:
Merge pull request #3031 from jmcmorris/more_pathing

More Entity Pathing

jmcmorris commented:
Still planning on cleaning things up per @zach2good feedback but was waiting for his response first :)

GodModeXI committed:
[core] add Sekkanoki functionality to mobskill_state.cpp  (#3033)

* core update : add Sekkanoki functionality to mobskill_state.cpp for trust utilization.

* update

* m_spentTP correction

* Update per change requests

* removing unused variable

claywar committed:
Refactor Tidal Talisman to use lookup table

claywar committed:
Merge pull request #3034 from LandSandBoat/tidal-talisman

Refactor Tidal Talisman to use lookup table

jmcmorris committed:
Removing entity ids from scripts and using offsets instead

jmcmorris commented:
Much like earlier, this PR seems to be stuck in some weird limbo. I'm going to try converting it into a draft and then back and see if that'll force it to detect the changes on my branch.

jmcmorris commented:
No luck. I'm not sure if it is Github or this PR. I'll give it until tomorrow morning and if it is still not updating then I'll remake the PR.

TeoTwawki commented:
> No luck. I'm not sure if it is Github or this PR. I'll give it until tomorrow morning and if it is still not updating then I'll remake the PR.

![Githubs having issues, hang tight](https://user-images.githubusercontent.com/6871475/197917130-833f8820-5af9-4661-9026-d56ad071205e.png)

TeoTwawki commented:
PR's are technically also issues you see - its the same backend
